src/realizations/catchment/Bmi_Multi_Formulation.cpp

```diff
@@ -385,10 +367,32 @@ double Bmi_Multi_Formulation::get_response(time_step_t t_index, time_step_t t_de
         }
         next_time_step_index++;
     }
-    // We know this safely ...
-    nested_module_ptr source = std::dynamic_pointer_cast<Bmi_Module_Formulation<bmi::Bmi>>(availableData[get_bmi_main_output_var()]);
-    // TODO: but we may need to add a convert step here in the case of mapping
-    return source->get_var_value_as_double(get_bmi_main_output_var());
+    // Find the right module for the main output, checking primary first
+    int index = get_index_for_primary_module();
+    std::vector<std::string> out_var_names = modules[index]->get_output_variable_names();
+    // If we don't find it there, look through the others
+    if (std::find(out_var_names.begin(), out_var_names.end(), get_bmi_main_output_var()) == out_var_names.end()) {
+        for (int i = 0; i < modules.size(); ++i) {
+            if (i == index) continue;
+            out_var_names = modules[i]->get_output_variable_names();
+            if (std::find(out_var_names.begin(), out_var_names.end(), get_bmi_main_output_var()) != out_var_names.end()) {
+                index = i;
+                break;
+            }
+        }
+    }
+    // With the module index, we can also pick the type
+    if (module_types[index] == "bmi_c") {
```

Comment:
Each of these formulation types is MACRO conditionally included, so these lines need appropriate macro guards for each one.